Add deprecation warnings for `send()`, `transfer()`, ABI coder v1, contract comparisons, virtual modifiers and `memory-safe-assembly` comment (#16174)
ethereum/solidity: docs/contracts/functions.rst

Reviewer: Perhaps not in this PR but I think we will have to go through the docs regarding send and transfer and see where we have to update them wrt best practices regarding send and transfer

Author: Yeah, I was thinking about this. We will need to update parts of the docs to reflect the deprecation of these functionalities.
I guess a separate PR for that would be better for organization purposes.

Reviewer: Yeah. It has to be done before we release though. :)

- Calling an external function which consumes a large amount of gas
- Sending Ether

.. warning::
    ``send()`` and ``transfer()`` are deprecated and scheduled for removal.
    See the section on :ref:`address` for more information.

    While it is possible to recreate the functionality by explicitly setting the limit to the value of the stipend (2300 gas),
    this value no longer holds its original meaning due to changing opcode costs.

>     It is recommended to use different means to protect against reentrancy.

Reviewer 2: This part should be moved to `address` docs as well:

```suggestion
    See the section on :ref:`address` for more information.
```

Author: Moved.

    you have to implement a receive Ether function (using payable fallback functions for receiving Ether is
    not recommended, since the fallback is invoked and would not fail for interface confusions
    on the part of the sender).

>     Note that ``send`` and ``transfer`` are scheduled to be deprecated in the next breaking version (0.9).

Reviewer: Don't we deprecate them sooner, with this PR, and then with 0.9 (or another future breaking release) they are removed? Also I think this could be its own warning block. Perhaps even the first one.

Author: I am not sure. I tried to find somewhere whether we define when it will be deprecated but couldn't locate anything yet.

Reviewer: Deprecated means it is available but it is encouraged to use something else (mostly because itll be removed in the future). So as soon as we merge the PR and have a release, it is deprecated. It was meant as a leading question. It should be phrased differently.

Author: I guess I was using deprecation instead of removal. I will change the wording here to reflect that.

    not recommended, since the fallback is invoked and would not fail for interface confusions
    on the part of the sender).

.. warning::

>     Note that ``send`` and ``transfer`` are deprecated and scheduled for removal in the next breaking version (0.9).

Reviewer: I still think this should be the first of the three warnings as it's probably the most relevant one. The text also heavily focuses on using `send` and `transfer` - which ideally should be changed to call with a link to the appropriate section in the documentation. This could be part of the follow-up? :)

Author: True. Moved it to be the first warning.
> This could be part of the follow-up? :)

I think these should be in the follow-up because we can then already have the compiler emitting the warnings while we adapt the docs for the breaking release.

    not recommended, since the fallback is invoked and would not fail for interface confusions
    on the part of the sender).

.. warning::

>     Note that ``send`` and ``transfer`` are scheduled to be deprecated in the next breaking version (0.9).

Reviewer: this is still wrong. we deprecate them now, they are scheduled for removal in v0.9.

Author: Done.

>     You are encouraged to use the function ``call`` with and optional desired amount of gas and

Reviewer: ```suggestion
    You are encouraged to use the ``call`` function with an optionally provided maximum amount of gas (default forwards all remaining gas) and
```

Reviewer: Also it would be cool to have a link to the documentation of ``call`` here.

Author: Done.

>     an empty parameter, e.g, ``call{value: amount}("")``.

Reviewer: ```suggestion
    an empty calldata parameter, e.g., ``call{value: amount}("")``.
```

Author: Done.

            // If someone sends Ether to that contract,
            // the transfer will fail, i.e. this returns false here.

>             // This will report a warning

Reviewer: ```suggestion
            // This will emit a deprecation warning
```

>             return testPayable.send(2 ether);

Reviewer: The new way of writing this without the gas limits would be this right?

```solidity
(bool sent,) = address(test).call{value: 2 ether}("");
return sent;
```

Author: Yes.

Reviewer: That should be part of the warning.
